### ISSUE TITLE
Angular Signals' 'this' context usage prevents passing the .set() function

### ISSUE THREAD

**Issue by chrisrocco** [angular/angular#52007]
### Which @angular/* package(s) are the source of the bug?

core

### Is this a regression?

No

### Description

The Signals library makes clever usage of the `this` context to avoid creating extra functions when creating new Signals.
```ts
signalFn.set = signalSetFn;   // 👈 Depends on 'this'
// vs
signalFn.set = signalSetFn.bind(node);
// or (fastest?)
signalFn.set = signalSetFn(node);  // 👈 Returns a bound closure
```

The problem is it's easy to lose the `this` context by using the function as a value.

```ts
const current = signal(0);
dataStream.subscribe(current.set);  // 👈 Wrong 'this' context leads to obscure errors
```

This surfaces a minor trade off between performance and correctness.
 
Microbenchmarks suggest optimal performance would be creating a closure like so -
```ts
const signalSetFn = (node) => (newValue) => ...;
signalFn.set = signalSetFn(node);  // 👈 In the factory function
```
^ https://measurethat.net/Benchmarks/Show/1328/0/bind-vs-closure-declaration

That's what I'd propose.

Alternatively, we should at least add a descriptive error message since this will come up a lot.

**Comment by pkozlowski-opensource** [comment]
> That's what I'd propose.

That's sounds reasonable to me. @chrisrocco would you like to crate a PR or do we want me to send a fix?

**Comment by alxhub** [comment]
Fixed by #51986!
